src/main/java/rx/internal/operators/OperatorConcat.java

```diff
@@ -179,6 +187,14 @@ void subscribeNext() {
                 }
             }
         }
+        
+        void produced(long c) {
+            arbiter.produced(c);
+            AtomicLong requestedField = requested;
+            if (requestedField.get() != Long.MAX_VALUE) {
+                requestedField.addAndGet(-c);
```

Comment:
I'd prefer to see correctness. Speculating on the probability of a race
condition seems like dangerous territory to be entering.

On Sun, 31 Jan 2016 21:24 David Karnok <notifications@github.com> wrote:

> In src/main/java/rx/internal/operators/OperatorConcat.java
> <https://github.com/ReactiveX/RxJava/pull/3589#discussion_r51358378>:
>
> > @@ -179,6 +187,14 @@ void subscribeNext() {
> >                  }
> >              }
> >          }
> > +
> > +        void produced(long c) {
> > +            arbiter.produced(c);
> > +            AtomicLong requestedField = requested;
> > +            if (requestedField.get() != Long.MAX_VALUE) {
> > +                requestedField.addAndGet(-c);
>
> In normal use, such change never happens; consumers either start with
> MAX_VALUE or some prefetch value, plus you have to be very precise to hit
> the narrow window between the two lines.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ReactiveX/RxJava/pull/3589/files#r51358378>.
>